### PR TITLE
Added support for custom snapshot writer

### LINKED ISSUE #36: Snapshot size issues
Hi

We've been using Eveneum as part of our event-sourced application for a few years now and so far it is serving us very well 👍
We would like to express our gratitude to @jkonecki once again, thank you !

## The issue with Snapshot sizes

During the past years, there has been a few issues, but there is one that keeps coming back regularly: the document size of Snapshots.
To be clear, this issue doesn't originate from Eveneum, but rather from Cosmos.
Cosmos has a document size limit of 2 MB (see https://learn.microsoft.com/en-us/azure/cosmos-db/concepts-limits#per-item-limits).

Generally, our snapshots are intended to be quite small, but occasionally a client will do something absurd and end up suddenly having a Snapshot which is larger than 2 MB. Often these clients can be convinced to change the way they work, but occasionally they have a valid (unforeseen) reason for their abnormal Snapshot size.

Whenever we encountered this issue in the past years, we've been changing our aggregates, streams or events to accommodate this limit. Sadly this often results in splitting a Stream that should not be split.
What makes it worse is that we might have thousands of streams with very small snapshots and 1 that hits the limit. Meaning 1 absurd stream, makes our whole domain worse.

## Idea: Store large Snapshots as BLOBs

Since this problem is very rare (in terms of number of streams), but is so disruptive, we've been thinking of a way to allow these large Snapshots as a special case. Of course, we cannot exceed the limit of Cosmos' document sizes, but we can store them in a different storage that does allow larger documents!

### Step-by-step
- When saving a Snapshot (before serialization) we'll check the size of the JSON that will be serialized.
  - If the size is smaller than X% of the document size limit (2 MB)
    - We persist it as we normally would.
  - If the size is larger
    - We will store the JSON as a BLOB in an [Azure Storage Account](https://learn.microsoft.com/en-us/azure/storage/common/storage-account-overview).
    - We then take the URI of the BLOB and store it (using Eveneum) in a Snapshot document that only contains this BLOB URI. This will be a very small document, which is useless without the linked BLOB.
- When querying a Stream
  - If we don't find a BLOB-snapshot-document
    - Then we continue as normal.
  - If we do find a BLOB-snapshot-document
    - We will replace the BLOB-Snapshot-document (read via Eveneum) with the actual Snapshot that was stored in BLOB storage.
    - _Note: if the BLOB would be unavailable, we could continue reading the stream to the next snapshot._

### ➕Positives
- Our domain is no longer impacted by these infrequent large Snapshots and can stay the way it should be

### ➖Negatives
- When persisting a Snapshot
  - We take a performance hit for checking the size of the Snapshot before persisting it
  - When dealing with a Snapshot that is deemed too large, we take a performance hit for storing it as a BLOB
- When querying a Snapshot that was deemed too large, we take a performance hit for loading it from the BLOB storage

## What does this have to do with Eveneum?

Nothing! The idea we described above can be fully implemented without a single change to Eveneum.

But...
1. We believe that this idea might be of value to other people using Eveneum.
2. We believe that one of the negatives, namely the performance hit for checking the size of the Snapshot before persiting it, could be reduced by changes to Eveneum. If we implement it as described above, then the Snapshot will be serialized twice (once in our code to check the size and [once in Eveneum](https://github.com/Eveneum/Eveneum/blob/master/Eveneum/EventStore.cs#L266) in order to persist it).

We were therefor wondering if there wouldn't be a way partially or fully include this idea into Eveneum?
The inclusion could go anywhere from:
- From (fully): Passing Storage Account credentials to Eveneum
- To (partially): Providing a fallback mechanism for the Snapshot persistance.

Aside from this, we were also just curious what you (@jkonecki) thought about this approach? Perhaps we're missing something? :)

Thank you in advance!

### PR DESCRIPTION
**Pull request by jkonecki**
Resolves #36

### CODE REVIEW

File: Eveneum.Tests/Eveneum.Tests.csproj [hunk 1]
```diff
@@ -5,9 +5,6 @@
   <ItemGroup>
     <Content Remove="C:\Users\Jakub Konecki\.nuget\packages\microsoft.azure.cosmos\3.15.0\contentFiles\any\netstandard2.0\ThirdPartyNotice.txt" />
```
Comment (bartelink) [inline Eveneum.Tests/Eveneum.Tests.csproj:6]: eh? consider updating cosmos sdk ver - the issue of the spurios content has long been resolved

**Comment by jkonecki** [comment]
Thank you, but I want to keep the minimum Cosmos SDK version as low as
possible.

Individual users can install higher versions of they wish.

**Comment by bartelink** [comment]
that's not unreasonable, though I'd consider:
- cosmos have a minimum viable recommended version; I'm pretty sure its newer than this
- if you are retaining the Remove, code it to not be hardwired to your user path

**Comment by jkonecki** [comment]
This is a brilliant suggestion! Upgrading to version 3.35.4 now.

Many thanks, Ruben.

________________________________

**Pull request merged**